common/core/desktop/tests/unit/kmnkbd/test_kmx_xstring.cpp

```diff
@@ -0,0 +1,230 @@
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000008\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3);
+
+  // --- Test for FFFF  CODE_EXTENDED -------------------------------------------------------------------------------------------------------- 
+  p = (PKMX_WCHAR)u"\U0000FFFF\U0000000A\U00000001\U00000002\U00000003\U00000004\U00000005\U00000006\U00000010";
+  q = incxstr(p);
+  assert(q == p + 9);
+
+  // --- Test for FFFF +CODE_CLEARCONTEXT ------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U0000000E\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3);
+
+  // --- Test for FFFF +CODE_CALL ----------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U0000000F\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3);
+
+  // --- Test for FFFF +CODE_CONTEXTEX ---------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000011\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3);
+
+  // --- Test for FFFF +CODE_IFOPT -------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000014\U00000002\U00000002\U00000001";
+  q = incxstr(p);
+  assert(q == p + 5);
+
+  // --- Test for FFFF +CODE_IFSYSTEMSTORE ------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000017\U00000002\U00000002\U00000001";
+  q = incxstr(p);
+  assert(q == p + 5);
+
+  // --- Test for FFFF +CODE_SETOPT ----------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000013\U00000002\U00000001";
+  q = incxstr(p);
+  assert(q == p + 4);
+
+  // --- Test for FFFF +CODE_SETSYSTEMRESTORE ---------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000018\U00000002\U00000001";
+  q = incxstr(p);
+  assert(q == p + 4);
+
+  // --- Test for FFFF +CODE_RESETOPT -----------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000016\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3);
+
+  // --- Test for FFFF +CODE_SAVEOPT -----------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000015\U00000001";
+  q = incxstr(p);
+  assert(q == p + 3 );
+
+  // --- Test for FFFF +default ----------------------------------------------------------------------------------------------------------------
+  p = (PKMX_WCHAR)u"\U0000FFFF\U00000004";
+  q = incxstr(p);
+  assert(q == p + 2);
+
```

Comment:
These tests are all good. It is important that we add some more normal case tests for longer strings, so we can be certain that `incxstr` is not simply terminating at the final nul, e.g.

```
p = (KMX_WCHAR)u"\U0000FFFF\U00000004\U00001234\U00002468";
q = incxstr(p);
assert(q == p + 2);
```

It might be good to duplicate all the tests to do this. Or you may want to create a helper function to simplify your work:

```
  test_incxstr_offset(u"\U0000FFFF\U00000004", 2);
```

```
void test_incxstr_offset(PKMX_WCHAR p, int offset) {
  PKMX_WCHAR q = incxstr(p);
  assert(q == p+ offset);
  //TODO: allocate buffer to copy p into a new string with extra chars \U00001234\U00002468 
  //on the end and then test that string also, then free buffer.
  //If you do go with a helper function, then it would be good to add some other small tests
  //here, e.g. do the same thing with a UC_SENTINEL or a surrogate pair,
}
```

Reply:
I wrote a helper function but then realized that special care would have been necessary to ensure that the concatenated sequences would be stored in the buffer in a way incxstr()could be tested. The helper function worked OK for most cases but with \0 it wasn`t behaving correctly:
When a \0 is copied to the buffer with wcscpy()  and then other characters like "\U00001234\U00002468" are concatenated using  wcscat() the \0 in the buffer would be removed and only those 2 characters would be placed in the buffer. This makes incxstr() move the pointer forward instead of stopping at (the lost) \0. Therefore, I decided to add more tests without using a helper function.